concordium-contracts-common/src/schema_json.rs

```diff
@@ -953,192 +1472,347 @@ fn deserial_string<R: Read>(source: &mut R, size_len: SizeLength) -> ParseResult
+                        json!({ "contract": name_without_init })
+                    })
+                    .map_err(|e| deserial_error(e.0))
             }
             Type::ReceiveName(size_len) => {
-                let owned_receive_name = OwnedReceiveName::new(deserial_string(source, *size_len)?)
-                    .map_err(|_| ParseError::default())?;
-                let receive_name = owned_receive_name.as_receive_name();
-                let contract_name = receive_name.contract_name();
-                let func_name = receive_name.entrypoint_name();
-                Ok(json!({"contract": contract_name, "func": func_name}))
+                let name = deserial_string(source, *size_len).map_err(|e| {
+                    ParseErrorWithReason(format!("Could not parse receive name from value ({})", e))
+                });
+                let owned_receive_name = name.and_then(|n| {
+                    OwnedReceiveName::new(n)
+                        .map_err(|e| ParseErrorWithReason(format!("Invalid receive name ({})", e)))
+                });
+
+                owned_receive_name
+                    .map(|orn| {
+                        let receive_name = orn.as_receive_name();
+                        let contract_name = receive_name.contract_name();
+                        let func_name = receive_name.entrypoint_name();
+                        json!({"contract": contract_name, "func": func_name})
+                    })
+                    .map_err(|e| deserial_error(e.0))
             }
             Type::ULeb128(constraint) => {
-                let int = deserial_biguint(source, *constraint)?;
+                let int = deserial_biguint(source, *constraint).map_err(|_| {
+                    deserial_error("Could not parse unsigned integer from value".into())
+                })?;
                 Ok(Value::String(int.to_string()))
             }
             Type::ILeb128(constraint) => {
-                let int = deserial_bigint(source, *constraint)?;
+                let int = deserial_bigint(source, *constraint).map_err(|_| {
+                    deserial_error("Could not parse signed integer from value".into())
+                })?;
                 Ok(Value::String(int.to_string()))
             }
             Type::ByteList(size_len) => {
-                let len = deserial_length(source, *size_len)?;
-                let mut string =
-                    String::with_capacity(std::cmp::min(MAX_PREALLOCATED_CAPACITY, 2 * len));
-                for _ in 0..len {
-                    let byte = source.read_u8()?;
-                    string.push_str(&format!("{:02x?}", byte));
+                let len = deserial_length(source, *size_len).map_err(|_| {
+                    ParseErrorWithReason("Could not parse ByteList length from value".into())
+                });
+                let bytes: core::result::Result<Vec<_>, _> =
+                    len.as_ref().map_err(|e| e.to_owned()).and_then(|len| {
```

Comment:
Why `to_owned` instead of `clone`?

Reply:
No reason, if you prefer `clone` I can do that instead. Care to explain why you  prefer one over the other though?

Reply:
Clone is more direct and I think what you want here since you are going from `&A` to `A`.

`ToOwned` is more generic in that it allows you to change types, e.g., go from `&str` to `String`.

Of course there is a generic implementation of ToOwned for any Clone https://doc.rust-lang.org/src/alloc/borrow.rs.html#82 which makes these equivalent in this case, but I think it is better to be direct rather than generic.

Reply:
Thanks for the elaborate response! 👍